README.md

```diff
@@ -138,6 +138,80 @@ android {
 }
 ```
 
+## @GenerateTypeAdapter
+
+There is an annotation in the `auto-value-gson-annotations` artifact called `@GenerateTypeAdapter`. This annotation
+can be set on types to indicate to the extension that you want the generated adapter to be a top level class in the same
+package. The name of this class will be the AutoValue class's name plus `_GsonTypeAdapter` suffix.
+
+Types annotated with this can also be (de)serialized dynamically at runtime with a provided runtime `TypeAdapterFactory`
+implementation in the annotation called `FACTORY`. The type name and generated typeadapter class's name *must not be obfuscated*
+for this to work.
+
+When this annotation is used, there will be no intermediate AutoValue class generated (as opposed to the default logic, 
+which generates an intermediate class and generates the `TypeAdapter` as a static inner class of it).
+
+`@GenerateTypeAdapter` is compatible with the factory approach above, just make your static method's implementation
+point to it. It can also be an alternative to it if you use the runtime factory, particularly if you 
+have a multimodule project and are willing to accept a small amount of (heavily cached) reflection.
+
+The generated class will have the same parameters as if it were the inner class. If it's generic, its constructor
+accepts a `Gson` instance and `TypeToken` of the generics. If it's not generic, it's just a `Gson` instance.
+
+Example usage:
+
+```java
+@GenerateTypeAdapter
+@AutoValue
+public class Foo {
+  // ...
+  public static TypeAdapter<Foo> typeAdapter(Gson gson) {
+    return new Foo_GsonTypeAdapter(gson);
+  }
+}
+
+// Generates
+public final class Foo_GsonTypeAdapter extends TypeAdapter<Foo> {
+  public Foo_GsonTypeAdapter(Gson gson) {
+    //...
+  }
+}
+
+// Or with generics
+@GenerateTypeAdapter
+@AutoValue
+public class Foo<T> {
+  // ...
+  public static TypeAdapter<Foo> typeAdapter(Gson gson, TypeToken<? extends Foo<T>> typeToken) {
+    return new Foo_GsonTypeAdapter(gson, typeToken);
+  }
+}
+
+// Generates
+public final class Foo_GsonTypeAdapter extends TypeAdapter<Foo> {
+  public Foo_GsonTypeAdapter(Gson gson, TypeToken<? extends Foo<T>> typeToken) {
+    //...
+  }
+}
+
+// Using the runtime FACTORY
+new GsonBuilder()
+    .addTypeAdapterFactory(GenerateTypeAdapter.FACTORY)
+    .build()
+    .toJson(myFooInstance);
+```
+
+Proguard rules for obfuscation:
+
+```
+# Retain generated classes that end in the suffix
```

Comment:
I _think_ this works, but going to ask a couple proguard-knowledgable friends to double check. We should only need to keep the names

Reply:
I'm curious if this is even necessary.  This is a simple copy/paste, since the wildcards mean that nothing changes based on the implementation, and we now have a compile scoped artifact, so can we include the proguard config in the artifact the same way other libraries do?

Reply:
Turns out the answer is no.  From Eric Lafortune:

> ProGuard consumer files are part of the aar specification, and implemented in the Android Gradle plugin (and the DexGuard Gradle plugin). I don't think there's a similar specification or implementation in the Java world.

So unless we make the dependency an aar, which I don't think we want to do, this is the way to go.

Reply:
Yep. Agreed it doesn't make sense to make this an AAR as well.